### PR TITLE
Bulk executor impl fixes for tests

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
To pass tests to enable bulk execution backend (https://github.com/ray-project/ray/pull/30903).

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_dataset.py [hunk 1]
```diff
@@ -5414,13 +5414,17 @@ def f(x):
...
-        num_cpus * (1 / compute_strategy.ready_to_total_workers_ratio)
-    )
-    assert (
-        compute_strategy.num_workers >= num_cpus
-        and compute_strategy.num_workers <= expected_max_num_workers
-    ), "Number of actors is out of the expected bound"
+
+    # The new execution backend is not using the ActorPoolStrategy under
+    # the hood, so the expectation here applies only to the old backend.
+    if not DatasetContext.get_current().new_execution_backend:
```
Comment (ericl) [inline python/ray/data/tests/test_dataset.py:5420]: Could you add a TODO for @clarkzinzow to remove this after the actor pool autoscaling fix?

**Review by ericl — APPROVED** [review]

File: python/ray/data/tests/test_dataset.py (hunk 1 above)
Comment (c21) [reply python/ray/data/tests/test_dataset.py:5420]: Let's file an issue for better tracking. Thanks.

**Review by c21 — APPROVED** [review]

File: python/ray/data/tests/test_dataset.py (hunk 1 above)
Comment (ericl) [reply python/ray/data/tests/test_dataset.py:5420]: We could add it to the acceptance criteria for https://github.com/ray-project/ray/issues/31723

**Review by clarkzinzow — APPROVED** [review]
LGTM!

File: python/ray/data/tests/test_dataset.py (hunk 1 above)
Comment (jianoaix) [reply python/ray/data/tests/test_dataset.py:5422]: Done!

**Pull request merged**